### ISSUE TITLE
MuPDF doesn't provide mupdf-third anymore.

### ISSUE THREAD

**Issue by mrpingo** [milgra/mmfm#2]
This project is awesome, congrats!
Although app-text/mupdf does not provide mupdf-thrid anymore. So cannot compile in Gentoo Linux. Is there anyway to overcome this issue? Can lightweight pdf reader like Zathura serve the same purspose?

❯ meson build --buildtype=release
The Meson build system
Version: 0.63.3
Source dir: /home/mrpingo/gits/mmfm/mmfm
Build dir: /home/mrpingo/gits/mmfm/mmfm/build
Build type: native build
Project name: mmfm
Project version: 0.43b
C compiler for the host machine: ccache cc (gcc 12.2.0 "cc (Gentoo 12.2.0 p1) 12.2.0")
C linker for the host machine: cc ld.bfd 2.39
Host machine cpu family: x86_64
Host machine cpu: x86_64
Library rt found: YES
Library m found: YES
Library mupdf found: YES

meson.build:19:0: ERROR: C shared or static library 'mupdf-third' not found

**Comment by milgra** [comment]
Hi, thank you for trying! On debian mupdf-dev contains mupdf-third, I don't know the state of it on gentoo, in the worst case you have to compile it for yourself.

**Comment by mrpingo** [comment]
I meant that over version 1.18, mupdf no longer provides libmupdf-third.so (third party sources built into libmupdf.so). Are you using an old version of it? I have installed mupdf-1.20.3

Reference:
https://www.mail-archive.com/pld-cvs-commit@lists.pld-linux.org/msg475799.html

**Comment by rik1599** [comment]
Same problem on arch-linux

**Comment by trinitronx** [comment]
> Same problem on arch-linux

I can confirm that on Arch-based distros (all rolling release), we no longer have `libmupdf-third.a` static library.  I've added some [comments on the AUR package here][1] to leave some breadcrumbs for those using that package, since I don't have access to the AUR `PKGBUILD` git repo.  [I pushed some `PKGBUILD` changes to this fork & branch here][2], which fixes the missing `gumbo-parser` dependency and updates to `mmfm` version `0.53b`.  However, I also wasn't sure about what the state of `libmupdf-third` was until digging into this further.  It looks like the problem is [in `meson.build` adding a dependency on `libmupdf-third`][2.1] in this upstream repo.

On Arch-based distros... There are now only `libmupdf.so` and `libmupdf.so` with versioned symlinks for build-time linkage & dynamic linking (e.g. [ldconfig](https://man7.org/linux/man-pages/man8/ldconfig.8.html) + [ld-linux.so](https://man7.org/linux/man-pages/man8/ld.so.8.html) and friends).  **`libmupdf-third.a` no longer exists**:

```console
$ cat /etc/os-release
NAME="Manjaro Linux"
PRETTY_NAME="Manjaro Linux"
ID=manjaro
ID_LIKE=arch
BUILD_ID=rolling
ANSI_COLOR="32;1;24;144;200"
HOME_URL="https://manjaro.org/"
DOCUMENTATION_URL="https://wiki.manjaro.org/"
SUPPORT_URL="https://forum.manjaro.org/"
BUG_REPORT_URL="https://docs.manjaro.org/reporting-bugs/"
PRIVACY_POLICY_URL="https://manjaro.org/privacy-policy/"
LOGO=manjarolinux

$ pacman -Ql libmupdf | grep -iE '\.so|\.a'
libmupdf /usr/lib/libmupdf.so
libmupdf /usr/lib/libmupdf.so.25.2
libmupdf /usr/lib/libmupdf.so
libmupdf /usr/lib/libmupdfcpp.so.25.2
```

> On debian mupdf-dev contains mupdf-third [...]

Looks like this is now removed in Debian Sid.  We now only have `libmupdf.so` and `libmupdfcpp.so`:

```console
$ docker run -ti --rm debian:sid /bin/bash

root@c15cddbbe37b:/# apt-get update -qq
root@c15cddbbe37b:/# apt-get install -y libmupdf-dev > /dev/null 
debconf: delaying package configuration, since apt-utils is not installed

root@c15cddbbe37b:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a'
/usr/lib/x86_64-linux-gnu/libmupdf.so
/usr/lib/x86_64-linux-gnu/libmupdfcpp.so
```

In Debian `12` / Bookworm, we still have the static libraries `libmupdf-third.a` and `libmupdf.a`:

<details><summary>Expand for Details</summary>
<p>

```console
$ docker run -ti --rm debian:12 /bin/bash

root@8e7c8d9988a7:/# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

root@8e7c8d9988a7:/# apt-get update -qq
root@8e7c8d9988a7:/# apt-get install -y libmupdf-dev > /dev/null 
debconf: delaying package configuration, since apt-utils is not installed

root@8e7c8d9988a7:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a'
/usr/lib/libmupdf-third.a
/usr/lib/libmupdf.a
```

</p>
</details> 

In Debian `11` / Bullseye, there is only `libmupdf.a` static library:

<details><summary>Expand for Details</summary>
<p>

```console
$ docker run -ti --rm  debian:11 /bin/bash                                                                                                      
root@7dbf92711669:/# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

root@7dbf92711669:/# apt-get update -qq 
root@7dbf92711669:/# apt-get install -y libmupdf-dev > /dev/null
debconf: delaying package configuration, since apt-utils is not installed

root@7dbf92711669:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a'
/usr/lib/libmupdf.a
```

</p>
</details> 

In Debian `10` / Buster, we again have `libmupdf-third.a` and `libmupdf.a` static libs:

**Note:** [Debian `10` / Buster was EoL as of June 30th, 2024.][3]

<details><summary>Expand for Details</summary>
<p>

```console
$ docker run -ti --rm debian:10 /bin/bash

root@a3be8bff394d:/# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

root@a3be8bff394d:/# apt-get update -qq
root@a3be8bff394d:/# apt-get install -y libmupdf-dev > /dev/null 
debconf: delaying package configuration, since apt-utils is not installed

root@a3be8bff394d:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a'
/usr/lib/libmupdf-third.a
/usr/lib/libmupdf.a
```

</p>
</details>

Similar story for Ubuntu releases, except we still have the static libraries including `libmupdf-third.a` up until <code>24.10</code> LTS (Oracular Oriole).

Only the newest Ubuntu `25.04` (Plucky Puffin) development branch has updated to have `libmupdf.so` and `libmupdfcpp.so` dynamic libs:

<details><summary>Expand for Ubuntu <code>25.04</code> LTS (Plucky Puffin) Details</summary>
<p>

```console
$ docker run -ti --rm ubuntu:25.04 /bin/bash

Digest: sha256:008b026f11c0b5653d564d0c9877a116770f06dfbdb36ca75c46fd593d863cbc
Status: Downloaded newer image for ubuntu:25.04

root@638fda1f2641:/# cat /etc/os-release 
PRETTY_NAME="Ubuntu Plucky Puffin (development branch)"
NAME="Ubuntu"
VERSION_ID="25.04"
VERSION="25.04 (Plucky Puffin)"
VERSION_CODENAME=plucky
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=plucky
LOGO=ubuntu-logo

root@638fda1f2641:/# apt-get update -qq
root@638fda1f2641:/# apt-get install -y libmupdf-dev > /dev/null 
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 79, <STDIN> line 39.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (Can't locate Term/ReadLine.pm in @INC (you may need to install the Term::ReadLine module) (@INC entries checked: /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.40.0 /usr/local/share/perl/5.40.0 /usr/lib/x86_64-linux-gnu/perl5/5.40 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl-base /usr/lib/x86_64-linux-gnu/perl/5.40 /usr/share/perl/5.40 /usr/local/lib/site_perl) at /usr/share/perl5/Debconf/FrontEnd/Readline.pm line 8, <STDIN> line 39.)
debconf: falling back to frontend: Teletype

root@638fda1f2641:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a' 
/usr/lib/x86_64-linux-gnu/libmupdf.so
/usr/lib/x86_64-linux-gnu/libmupdfcpp.so
```

</p>
</details> 

All the older Ubuntu images still have the static libs including `libmupdf-third.a`:

<details><summary>Expand for Ubuntu <code>24.10</code> LTS (Oracular Oriole) Details</summary>
<p>

```console
$ docker run -ti --rm ubuntu:24.10 /bin/bash

root@a60da1b544fe:/# cat /etc/os-release 
PRETTY_NAME="Ubuntu Oracular Oriole (development branch)"
NAME="Ubuntu"
VERSION_ID="24.10"
VERSION="24.10 (Oracular Oriole)"
VERSION_CODENAME=oracular
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=oracular
LOGO=ubuntu-logo

root@a60da1b544fe:/# apt-get update -qq
root@a60da1b544fe:/# apt-get install -y libmupdf-dev > /dev/null 
debconf: delaying package configuration, since apt-utils is not installed

root@a60da1b544fe:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a'
/usr/lib/libmupdf-third.a
/usr/lib/libmupdf.a
```

</p>
</details> 

<details><summary>Expand for Ubuntu <code>24.04</code> LTS (Noble Numbat) Details</summary>
<p>

```console
$ docker run -ti --rm ubuntu:24.04 /bin/bash

root@06606dfe9818:/# cat /etc/os-release 
PRETTY_NAME="Ubuntu 24.04 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo

root@06606dfe9818:/# apt-get update -qq
root@06606dfe9818:/# apt-get install -y libmupdf-dev > /dev/null 
debconf: delaying package configuration, since apt-utils is not installed

root@06606dfe9818:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a'
/usr/lib/libmupdf-third.a
/usr/lib/libmupdf.a
```

</p>
</details> 

<details><summary>Expand for Ubuntu <code>22.04</code> LTS (Jammy Jellyfish) Details</summary>
<p>

```console
$ docker run -ti --rm ubuntu:22.04 /bin/bash
root@f1d01881de4b:/# cat /etc/os-release 
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

root@f1d01881de4b:/# apt-get update -qq
root@f1d01881de4b:/# apt-get install -y libmupdf-dev > /dev/null 
debconf: delaying package configuration, since apt-utils is not installed

root@f1d01881de4b:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a'
/usr/lib/libmupdf-third.a
/usr/lib/libmupdf.a
```

</p>
</details> 

<details><summary>Expand for Ubuntu <code>20.04</code> LTS (Focal Fossa) Details</summary>
<p>

```console
$ docker run -ti --rm ubuntu:20.04 /bin/bash

root@4849c19f98b3:/# cat /etc/os-release 
NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

root@4849c19f98b3:/# apt-get update -qq
root@4849c19f98b3:/#  apt-get install -y libmupdf-dev > /dev/null 
debconf: delaying package configuration, since apt-utils is not installed

root@4849c19f98b3:/# dpkg -L  libmupdf-dev | grep -iE '\.so|\.a'
/usr/lib/libmupdf-third.a
/usr/lib/libmupdf.a
```

</p>
</details> 

> [...] `app-text/mupdf` does not provide `mupdf-third` anymore. So cannot compile in Gentoo Linux.

I can also confirm this.

For Gentoo we have no `libmupdf-third.a` static library anymore.  Just `libmupdf.so` and no C++ library, `libmupdfcpp.so`:

<details><summary>Expand for Gentoo Details (<code>gentoo/stage3:t64-20250127</code>)</summary>
<p>

```console
$ docker run -ti --rm gentoo/stage3:latest  /bin/bash

f424fbdad51d / # emaint --auto  sync > /dev/null
f424fbdad51d / # emerge  --oneshot app-text/mupdf

[...SNIP...] ## 38 package builds later ...

## Install equery
f424fbdad51d / # emerge app-portage/gentoolkit

f424fbdad51d / # equery files --filter=obj,sym app-text/mupdf | grep -iE '\.so|\.a'
/usr/lib64/libmupdf.so
/usr/lib64/libmupdf.so.1.24.8

## Also check for pkg-config file while we're at it
f424fbdad51d / # equery files --filter=obj,sym app-text/mupdf | grep -iE '\.pc'    
/usr/lib64/pkgconfig/mupdf.pc
f424fbdad51d / # cat /usr/lib64/pkgconfig/mupdf.pc
prefix=/usr
exec_prefix=${prefix}
libdir=${exec_prefix}/lib64
includedir=${prefix}/include

Name: mupdf
Description: Library for rendering PDF documents
Requires: freetype2 libopenjp2 libcrypto
Version: 1.24.8
Libs: -L${libdir} -lmupdf
Libs.private: -lmupdf-js-none
Cflags: -I${includedir}
```

```console
## Finally, checking the latest "unstable" version (masked by default)
f424fbdad51d / # ACCEPT_KEYWORDS='~amd64' emerge  --oneshot   =app-text/mupdf-1.25.2

[...SNIP...] ## package built ...

f424fbdad51d / # equery files --filter=obj,sym app-text/mupdf | grep -iE '\.so|\.a'
/usr/lib64/libmupdf.so
/usr/lib64/libmupdf.so.1.25.2

## pkg-config for mupdf-1.25.2
f424fbdad51d / # cat /usr/lib64/pkgconfig/mupdf.pc
prefix=/usr
exec_prefix=${prefix}
libdir=${exec_prefix}/lib64
includedir=${prefix}/include

Name: mupdf
Description: Library for rendering PDF documents
Requires: freetype2 libopenjp2 libcrypto
Version: 1.25.2
Libs: -L${libdir} -lmupdf
Libs.private: -lmupdf-js-none
Cflags: -I${includedir}
```

</p>
</details>

It seems that on Gentoo at least the default `ebuild`s for both versions available don't install the C++ shared library.  I couldn't figure out how to enable this if it's a use flag or `make` args.

[1]: https://aur.archlinux.org/packages/mmfm#comment-981100
[2]: https://github.com/trinitronx/mmfm/compare/master...fix-pkgbuild
[2.1]: https://github.com/milgra/mmfm/blob/75e7131a92e9f1bf8286718bd2cc0f20ae5be27b/meson.build#L119-L121
[3]: https://wiki.debian.org/LTS/

**Comment by trinitronx** [comment]
# TL;DR

The short version of all the above is that:

- `libmupdf-third.a` static library is deprecated in latest `mupdf`
- Latest `mupdf` now always ships at least a single dynamic library: `libmupdf.so`
  - Depending on distro or release, they may or may not include the C++ lib: `libmupdfcpp.so`
- Some Debian-family distro releases may still have the old `libmupdf-third.a` static library, but this is eventually going away as evident in the latest versions of those distros.


## Recommended Fix

The `meson.build` in this repo should be changed so:

- It will only link against `libmupdf-third` if it exists on the system.
- If `libmupdf-third` library doesn't exist, then just link against `libmupdf` only.
